File: lib/KittyCGI/kitty.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#################################################################################
#																				#
# Автор:	Косьмина Ярослав													#
# Сайт:		http://darviarush.narod.ru											#
#																				#
# Среда:	http://php.net														#
#																				#
#################################################################################

# локаль
import sys
reload(sys)
sys.setdefaultencoding("utf-8")

import traceback

ref = None

import json, codecs, re


def frisky_kitty(req, *v, **kv):
	r = [chr(6), req]
	if len(kv): v.append(kv)
	if len(v): r.append(' '); r.append(json.dumps(v))
	print "".join(r)
	sys.stdout.flush()
	
def kitty(req, *v, **kv):
	frisky_kitty(chr(6)+req, *v, **kv)
	line = sys.stdin.readline()
	if line[0]=='{' or line[0]=='[': line = json.loads(line)
	return line

actions = {}

request = sys.stdin.readline()

while request != '':
	request = request.rstrip()
	
	action = actions.get(request)
	if not action:
		action = 'kitty_' + re.sub(r'[/.-]', '__', request)
	
		file = codecs.open(request, 'rb', 'utf8').read()
		file = re.sub(r'^', '\t', file, 0, re.M)
		file = ''.join(["def ", action, "():\n", file])
		glob = {"frisky_kitty": frisky_kitty, "kitty": kitty}
		loc = {}
		exec( file, glob, loc )
		
		actions[request] = action = loc[action]

	try:
		ref = action()
	except BaseException as e:
		sys.stderr.write( traceback.format_exc() )

	request = kitty("end", ref)
	ref = None

File: lib/KittyCGI/kitty.rb
#!/usr/bin/env ruby
# encoding: UTF-8

#################################################################################
#																				#
# Автор:	Косьмина Ярослав													#
# Сайт:		http://darviarush.narod.ru											#
#																				#
# Среда:	https://www.ruby-lang.org											#
#																				#
#################################################################################

require 'json'

class KittyClass

	def frisky_kitty(req, *param)
		r = [6.chr, req]
		if param.size != 0
			r << " "
			r << JSON.generate(param)
		end
		puts r.join("")
		STDOUT.flush()
	end
		
	def kitty(req, *param)
		frisky_kitty(6.chr << req, *param)
		line = STDIN.readline()
		if line[0] == '[' or line[0] == '{'
			line = JSON.parse(line)
		end
		line
	end
end


Kitty = KittyClass.new

actions = {}

request = STDIN.readline()

while request != ''
	request = request.chomp()
	
	action = actions[request]
	if not action
		action = 'kitty_' << request.gsub(/[\/.-]/, '__')
	
		file = File.read(request)
		file = ["class KittyClass\ndef ", action, "()\n", file, "\nend\nend\n"].join('')
		
		eval( file )
		
		actions[request] = action
	end

	begin
		ref = Kitty.send action
	rescue Exception => e
		$stderr.puts e.message
		for i in e.backtrace
			$stderr.puts i
		end
		
	end
	request = Kitty.kitty("end", ref)
	ref = nil
end
